### PR TITLE
Extract logic that calculates diagnostics for a problem into a shared service

### PR DESCRIPTION
**Pull request by adammurdoch**
Fixes #?

### Context

This PR moves some common logic shared by exception analysis, deprecation logging and configuration cache problem reporting into a shared service. Previously this logic was duplicated by each of the callers and behaved differently.

The service also includes some basic logic to limit the number of stack traces that are collected when many problems are encountered during build execution. In this PR, the limit is hardcoded to 50, regardless of the problem type. This is intentionally very simple and can be improved in later PRs.

This PR does not attempt to move all of the duplicated logic into the service. For example, stack trace sanitization and duplicate problem handling are still implemented in different ways for each type of problem. This duplication can be addressed in later PRs.

This PR also includes some refactoring to use the configuration cache `ProblemFactory` in more places so that the shared logic is used for more kinds of configuration cache problems. There are still some places that still create `PropertyProblem` instances directly that need to be refactored.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 1]
```diff
@@ -45,36 +47,39 @@ public class LoggingDeprecatedFeatureHandler implements FeatureHandler<Deprecate
...
     }
 
     @Override
-    public void featureUsed(DeprecatedFeatureUsage usage) {
+    public void featureUsed(final DeprecatedFeatureUsage usage) {
         deprecationsFound = true;
+        ProblemDiagnostics diagnostics = problemDiagnosticsFactory.forCurrentCaller(new ProblemDiagnosticsFactory.StackTraceTransformer() {
+            @Override
+            public List<StackTraceElement> transform(StackTraceElement[] original) {
+                return FeatureUsage.calculateStack(usage.getCalledFrom(), original);
```
Comment (wolfs) [inline subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:68]: :thinking: Should this be an actual class where the `calculateStack()` method is as well? The message seems somewhat orphaned on `FeatureUsage` now.

File: subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 2]
```diff
@@ -45,36 +47,39 @@ public class LoggingDeprecatedFeatureHandler implements FeatureHandler<Deprecate
...
-            locationReporter.reportLocation(usage, message);
-            if (message.length() > 0) {
+            Location location = diagnostics.getLocation();
+            if (location != null) {
+                message.append(location.getFormatted());
                 message.append(SystemProperties.getInstance().getLineSeparator());
             }
             message.append(featureMessage);
-            displayDeprecationIfSameMessageNotDisplayedBefore(message, usage.getStack());
+            displayDeprecationIfSameMessageNotDisplayedBefore(message, diagnostics.getStack());
```
Comment (wolfs) [inline subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:80]: :question: May this cause the message to be displayed an additional time as before, when we have a lot of deprecations with the same stack trace? We'd show it once with the (possible truncated) stacktrace and once without, right? I suppose that isn't a big deal.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt [hunk 3]
```diff
@@ -3037,7 +3037,10 @@ Class <org.gradle.internal.os.OperatingSystem$Solaris> is not annotated (directl
 Class <org.gradle.internal.os.OperatingSystem$Unix> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (OperatingSystem.java:0)
 Class <org.gradle.internal.os.OperatingSystem$Windows> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (OperatingSystem.java:0)
 Class <org.gradle.internal.os.OperatingSystem> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (OperatingSystem.java:0)
+Class <org.gradle.internal.problems.DefaultProblemDiagnosticsFactory$DefaultProblemDiagnostics> is not annotated (directly or via its package) with @org.gradle.api.NonNullApi in (DefaultProblemDiagnosticsFactory.java:0)
```
Comment (wolfs) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:3040]: 🤔 Can you add a `package-info.java` with `@NonNullApi` to the new package?

**Comment by wolfs** [comment]
This PR gives back the performance regression introduced when adding the `BuildIdentifier.buildPath` deprecation: 
image

**Review by wolfs — COMMENTED** [review]
The overall PR looks good. I left some comments.

File: subprojects/architecture-test/src/changes/accepted-public-api-changes.json [hunk 4]
```diff
@@ -1,3 +1,12 @@
 {
-    "acceptedApiChanges": []
-}
\ No newline at end of file
+    "acceptedApiChanges": [
+        {
+            "type": "org.gradle.util.SingleMessageLogger",
```
[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/architecture-test/src/changes/accepted-public-api-changes.json:4]: I think so.

File: subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 2 above)
Comment (adammurdoch) [reply subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:80]: Good point. I'll fix that.

File: subprojects/logging/src/test/groovy/org/gradle/internal/deprecation/DeprecationLoggerTest.groovy [hunk 5]
```diff
@@ -54,6 +59,7 @@ class DeprecationLoggerTest extends ConcurrentSpec {
 
         when:
         DeprecationLogger.reset()
+        DeprecationLogger.init(diagnosticsFactory, WarningMode.All, Mock(BuildOperationProgressEventEmitter))
```
[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/logging/src/test/groovy/org/gradle/internal/deprecation/DeprecationLoggerTest.groovy:62]: `reset()` throws away the diagnostics factory, so the test won't use the mock. On the other hand, it probably doesn't matter for that particular test.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt (hunk 3 above)
Comment (adammurdoch) [reply subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:3040]: Will do. I thought I had but I guess I missed that package.

File: subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java [hunk 6]
```diff
@@ -83,8 +83,8 @@ protected void configure(ServiceRegistration registration, List<PluginServiceReg
         registration.add(TaskPathProjectEvaluator.class);
         registration.add(DefaultFeatureFlags.class);
         registration.add(DefaultProblemLocationAnalyzer.class);
+        registration.add(DefaultProblemDiagnosticsFactory.class);
```
[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java:86]: This was intentional. I did consider applying the limit to each kind of problem but decided to keep it very, very simple in this first PR.

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 1 above)
Comment (adammurdoch) [reply subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java]: This is in a separate class now.

File: subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 2 above)
Comment (adammurdoch) [reply subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java]: Done

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt (hunk 3 above)
Comment (adammurdoch) [reply subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt]: Done

File: subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java:86]: I'll do this change in a follow up  PR.

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheUnsupportedTypesIntegrationTest.groovy [hunk 7]
```diff
@@ -134,7 +134,7 @@ class ConfigurationCacheUnsupportedTypesIntegrationTest extends AbstractConfigur
         problems.assertResultHasProblems(result) {
             withTotalProblemsCount(1)
             withUniqueProblems(
-                "Class `Foo`: accessing non-serializable type '${serviceType.name}'"
+                "Build file 'build.gradle': line 9: accessing non-serializable type '${serviceType.name}'"
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheUnsupportedTypesIntegrationTest.groovy:137]: Nice! 🎉

**Review by bamboo — COMMENTED** [review]
🎉

**Pull request merged**